controllers/devbox/internal/controller/devbox_controller.go

```diff
@@ -439,6 +415,255 @@ func (r *DevboxReconciler) getRuntime(ctx context.Context, devbox *devboxv1alpha
+			},
+		},
+	}
+
+	ingressSpec := networkingv1.IngressSpec{
+		IngressClassName: &className,
+		Rules: []networkingv1.IngressRule{
+			{
+				Host: devbox.Name + "." + r.HostPath,
+				IngressRuleValue: networkingv1.IngressRuleValue{
+					HTTP: &networkingv1.HTTPIngressRuleValue{
+						Paths: ingressPath,
+					},
+				},
+			},
+		},
+	}
+
+	wsIngress := &networkingv1.Ingress{
+		ObjectMeta: metav1.ObjectMeta{
+			Name:      devbox.Name + "-proxy-ingress",
+			Namespace: devbox.Namespace,
+		},
+		Spec: ingressSpec,
+	}
+
+	if _, err := controllerutil.CreateOrUpdate(ctx, r.Client, wsIngress, func() error {
+		return controllerutil.SetControllerReference(devbox, wsIngress, r.Scheme)
+	}); err != nil {
+		return err
+	}
+	return nil
+}
+
+func (r *DevboxReconciler) syncProxySvc(ctx context.Context, devbox *devboxv1alpha1.Devbox, recLabels map[string]string, servicePorts []corev1.ServicePort) error {
+	runtimecr, err := r.getRuntime(ctx, devbox)
+	if err != nil {
+		return err
+	}
+	servicePort := []corev1.ServicePort{
+		{
+			Name:       "devbox-ssh-port",
+			Port:       8080,
+			TargetPort: intstr.FromInt32(8080),
+			Protocol:   corev1.ProtocolTCP,
+		},
+	}
+	expectServiceSpec := corev1.ServiceSpec{
+		Selector: helper.GenerateProxyPodLabels(devbox, runtimecr),
+		Type:     corev1.ServiceTypeClusterIP,
+		Ports:    servicePort,
+	}
+	proxySvc := &corev1.Service{
+		ObjectMeta: metav1.ObjectMeta{
+			Name:      devbox.Name + "-proxy-svc",
+			Namespace: devbox.Namespace,
+			Labels:    helper.GenerateProxyPodLabels(devbox, runtimecr),
+		},
+	}
+	if _, err := controllerutil.CreateOrUpdate(ctx, r.Client, proxySvc, func() error {
+		// only update some specific fields
+		proxySvc.Spec.Selector = expectServiceSpec.Selector
+		proxySvc.Spec.Type = expectServiceSpec.Type
+		if len(proxySvc.Spec.Ports) == 0 {
+			proxySvc.Spec.Ports = expectServiceSpec.Ports
+		} else {
+			proxySvc.Spec.Ports[0].Name = expectServiceSpec.Ports[0].Name
+			proxySvc.Spec.Ports[0].Port = expectServiceSpec.Ports[0].Port
+			proxySvc.Spec.Ports[0].TargetPort = expectServiceSpec.Ports[0].TargetPort
+			proxySvc.Spec.Ports[0].Protocol = expectServiceSpec.Ports[0].Protocol
+		}
+		return controllerutil.SetControllerReference(devbox, proxySvc, r.Scheme)
+	}); err != nil {
+		return err
+	}
+	return nil
+}
+
+func (r *DevboxReconciler) syncProxyPod(ctx context.Context, devbox *devboxv1alpha1.Devbox, recLabels map[string]string, servicePorts []corev1.ServicePort) error {
+	runtimecr, err := r.getRuntime(ctx, devbox)
+	if err != nil {
+		return err
+	}
+
+	sshPort := "22"
+	for _, port := range servicePorts {
+		if port.Name == "devbox-ssh-port" {
+			sshPort = port.TargetPort.String()
+			break
+		}
+	}
+
+	podName := devbox.Name + "-proxy-pod"
+	wsPod := &corev1.Pod{}
+	err = r.Client.Get(ctx, types.NamespacedName{Name: podName, Namespace: devbox.Namespace}, wsPod)
+	if err == nil {
+		return nil
+	}
+	if !errors.IsNotFound(err) {
+		return err
+	}
+	wsPod = &corev1.Pod{
```

Comment:
maybe use deployment would be better when handling exited pod.

Reply:
already fix it.